CONTRIBUTING.md

```diff
@@ -3,16 +3,53 @@
+
+# Reconcile Loop Design Principles
+
+* Never block reconcile execution. ever. This means __no sleeps and no retry loops__
+* All __logic must be idempotent__. This means discovering what has already occurred by inspecting informer cache, and only mutating to converge on desired state.
+* __Never perform an apiserver List()__ request within the reconcile loop. This action is very apiserver intensive. Instead use informers retrieve items from cache.
+* __Avoid mutating a cluster object the loop didn't create__. A exception to this would to be adding annotations to an existing object. Mutating a object's Spec that our controller didn't create should be avoided.
+* __Avoid performing an apiserver Get()__ request within the reconcile loop if at all possible. Again use informers. There are rare exceptions.
+* Use __finalizers__ in order to perform any cleanup tasks required before a CR is completely removed from etcd.
+* User __owner references__ on any objects created and owned by a CR. Owner references ensure objects are automatically garbage collected after a corresponding CR is deleted.
+* Return early to handle errors which re-enqueues the work queue key to be tried again.
+* Do not create logic to sync a key using a recurring time interval. Instead only wake up processing the key when something has changed.
+* Reconcile execution should be broken into two parts, synchronization and status reporting. Synchronization converges the CR on the desired state. Status reporting records the current state of the CR based on the synchronization execution results and the observed state of the cluster. These are two independent functions. __Do not modify status during the synchronization step__.
+
+# API Design
+
+* All workflows __must be declarative__. This means the cluster-admin must be able to declare the exact state they want up front, and have our operator handle converging on that state.
+* __No imperative actions__. This means no requirement for a cluster-admin to account for the ordering or timing at which manifests are posted. If our API can't express the desired state up front, then the API is wrong.
+* Limit the Status.Phase enum to only express the state in broad terms. Use events and conditions on the CR to express detailed error information.
+* __Avoid design decisions that result in mutating objects not created and owned by our controller__. Mutating the spec of an object our controller doesn't own is nearly never an acceptable decision. However, there are occasions where adding annotations or labels to an object not owned by the controller makes sense.
+
+# Unit Testing
+
+Unit testing allows us to validate our reconcile loops under specific conditions to ensure no unexpected changes in behavior are introduced.
+
+* Use mock clients to simulate reconcile loops.
+* Execute reconcile loop to validate execution under specific conditions.
+* Validate reconcile execution by introspecting API calls that were made during the execution.
```

Comment:
I would prefer a link to a cook book example here myself.

Reply:
good idea. I added an example of a test using mock clients